html/user/download.php

```diff
@@ -16,110 +16,214 @@
+            return 'windows_x86_64';
+        } else {
+            return 'windows_intelx86';
+        }
+    } else if (strstr($client_info, 'Mac')) {
+        if (strstr($client_info, 'PPC Mac OS X')) {
+            return 'powerpc-apple-darwin';
+        } else {
+            return 'x86_64-apple-darwin';
+        }
+    } else if (strstr($client_info, 'Android')) {
+        // Check for Android before Linux,
+        // since Android contains the Linux kernel and the
+        // web browser user agent string lists Linux too.
+        //
+        return 'arm-android-linux-gnu';
+    } else if (strstr($client_info, 'Linux')) {
+        if (strstr($client_info, 'x86_64')) {
+            return 'x86_64-pc-linux-gnu';
+        } else {
+            return 'i686-pc-linux-gnu';
+        }
+    } else {
+        return null;
     }
-    if ($download) {
-        $x .= "&download=1";
-        if (parse_bool(get_config(), "need_vbox")) {
-            $x .= "&need_vbox=1";
+}
+
+// find release version for user's platform
+//
+function get_version() {
+    $v = simplexml_load_file("versions.xml");
+    $client_info = $_SERVER['HTTP_USER_AGENT'];
+    $p = client_info_to_platform($client_info);
+    foreach ($v->version as $i=>$v) {
+        if ((string)$v->dbplatform != $p) {
+            continue;
+        }
+        if (!strstr((string)$v->description, "ecommended")) {
+            continue;
         }
+        return $v;
     }
-    return $x;
+    return null;
 }
 
-function show_download_page() {
-    page_head("Download required software");
+function download_button($v, $project_id, $token, $user) {
+    return sprintf(
+        '<form action="https://boinc.berkeley.edu/concierge.php" method="post">
+        <input type=hidden name=project_id value="%d">
+        <input type=hidden name=token value="%s">
+        <input type=hidden name=user_id value="%d">
+        <input type=hidden name=filename value="%s">
+        <button class="btn btn-info">
+        <font size=2><u>Download BOINC</u></font>
+        <br>for %s (%s MB)
+        <br><small>BOINC %s</small></button>
+        </form>
+        ',
+        $project_id,
+        $token,
+        $user->id,
+        (string)$v->filename,
+        (string)$v->platform,
+        (string)$v->size_mb,
+        (string)$v->version_num
+    );
+}
+
+function download_button_vbox($v, $project_id, $token, $user) {
+    return sprintf(
+        '<form action="https://boinc.berkeley.edu/concierge.php" method="post">
+        <input type=hidden name=project_id value="%d">
+        <input type=hidden name=token value="%s">
+        <input type=hidden name=filename value="%s">
+        <button class="btn btn-success">
+        <font size=+1><u>Download BOINC + VirtualBox</u></font>
+        <br>for %s (%s MB)
+        <br><small>BOINC %s, VirtualBox %s</small></a>
+        </form>
+        ',
+        $project_id,
+        $token,
+        $user->id,
```

Comment:
There is no `user_id` field in the HTML above.

Reply:
I opened issue #2224  for this since this pull request has been merged (apparently prematurely - sorry about that).